packages/plugin-ext/src/main/browser/menus/menus-contribution-handler.ts

```diff
@@ -45,10 +46,16 @@ export class MenusContributionPointHandler {
                 const menus = contributions.menus[location];
                 menus.forEach(menu => {
                     const [group = '', order = undefined] = (menu.group || '').split('@');
-                    this.menuRegistry.registerMenuAction([...menuPath, group], {
-                        commandId: menu.command,
-                        order
-                    });
+                    // Registering a menu action requires the related command to be already registered.
+                    // But the contributed commands are registered after the menus.
+                    // So, before registering a menu action need to wait ~2 sec. while all the contributed commands be registered.
+                    // Workaround for https://github.com/theia-ide/theia/issues/3344
+                    setTimeout(() => {
+                        this.menuRegistry.registerMenuAction([...menuPath, group], {
+                            commandId: menu.command,
+                            order
+                        });
+                    }, 2000);
```

Comment:
First time commenting on your work on plugins, but wouldn't it be better to have some kind of event emitter/promise that would fire/resolve once the commands are ready, instead of hardcoding a timeout waiting for 2 seconds?

If the current way works with you guys then I don't mind :)

Reply:
Hello @marechal-p 
Thank you for pointing that out. I understand your concern. So let me explain a little bit.
Actually, the problem is related to different ways of registering the commands by Theia and VSCode plugins.

_VSCode plugins_
A plugin registers a menu action and a related command (id, title) via the `menus` and `commands` contribution points. Theia reads the `package.json` and registers all the contributed commands and then the menu actions.

_Theia plugins_
A plugin registers a menu action in the same way as a VSCode plugin. But a command is registered dynamically from the code using the Commands API (`theia.commands.registerCommand()`). And Theia doesn't know when all the contributed commands are registered.

Registering the commands by Theia plugin should work the same way as for VSCode plugin. There's the related issue already registered https://github.com/theia-ide/theia/issues/3344
Since it hasn't been fixed yet I need a temporary workaround.

Reply:
> Since it hasn't been fixed yet I need a temporary workaround.

I understand, it is fine :)

> But a command is registered dynamically from the code using the Commands API (`theia.commands.registerCommand()`). And Theia doesn't know when all the contributed commands are registered.

This is why we have specific steps in the frontend application, such as `MenuContribution.registerMenus()` which essentially is called from within `FrontendApplicationContribution.onStart()`. No async registration allowed AFAIK.

I was thinking that it would make sense to have a registration phase where you know that at some point everyone should be done with its contributions.

Just food for thoughts.

Please don't wait to address this comment before merging, then again just fix your issue :)